### PR TITLE
[clangtidy] Allow safe suspensions in coroutine-hostile-raii check

### PR DESCRIPTION
**Pull request by usx95**
Certain `awaitable` types could be safe to `co_await` on even when we have suspension-hostile RAII objects in scope. 
This PR adds a way for users to mark such safe `awaitable` and silence false positive warnings in `co_await` expressions involving such an `awaitable`.

`co_await`-ing an expression of `awaitable` type is considered safe if the type is part of `SafeAwaiatablesList` check option. RAII objects persisting across such a `co_await` expression are
considered safe and hence are not flagged.

### CODE REVIEW

**Comment by PiotrZSL** [comment]
Maybe instead of "coro_raii_safe_suspend" we should just use an list of "allowed types" ?

**Comment by usx95** [comment]
That is a good solution and does not involve changing the code. Done.

File: clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp [hunk 1]
```diff
@@ -52,27 +52,41 @@ AST_MATCHER_P(Stmt, forEachPrevStmt, ast_matchers::internal::Matcher<Stmt>,
...
+// Matches the expression awaited by the `co_await`.
+AST_MATCHER_P(CoawaitExpr, awaiatable, ast_matchers::internal::Matcher<Expr>,
+              InnerMatcher) {
+  return Node.getCommonExpr() &&
+         InnerMatcher.matches(*Node.getCommonExpr(), Finder, Builder);
+}
+
+auto typeWithNameIn(const std::vector<StringRef> &Names) {
+  return hasType(
+      hasCanonicalType(hasDeclaration(namedDecl(hasAnyName(Names)))));
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp:65]: There is other matcher in utils directory used by other checks that enable regexp.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp:65]: But thats also fine.

File: clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp [hunk 2]
```diff
@@ -52,27 +52,41 @@ AST_MATCHER_P(Stmt, forEachPrevStmt, ast_matchers::internal::Matcher<Stmt>,
   }
   return IsHostile;
 }
+
+// Matches the expression awaited by the `co_await`.
+AST_MATCHER_P(CoawaitExpr, awaiatable, ast_matchers::internal::Matcher<Expr>,
+              InnerMatcher) {
+  return Node.getCommonExpr() &&
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp:59]: avoid calling getCommonExpr twice, just put it in if or something

**Review by PiotrZSL — APPROVED** [review]
Overall looks +- fine for me.
Wait like 1-2 weeks before merging if someone got some more comments.

File: clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp [hunk 3]
```diff
@@ -52,27 +52,42 @@ AST_MATCHER_P(Stmt, forEachPrevStmt, ast_matchers::internal::Matcher<Stmt>,
   }
   return IsHostile;
 }
+
+// Matches the expression awaited by the `co_await`.
+AST_MATCHER_P(CoawaitExpr, awaiatable, ast_matchers::internal::Matcher<Expr>,
```
Comment (bjosv) [inline clang-tools-extra/clang-tidy/misc/CoroutineHostileRAIICheck.cpp:57]:
```suggestion
AST_MATCHER_P(CoawaitExpr, awaitable, ast_matchers::internal::Matcher<Expr>,
```

**Review by bjosv — COMMENTED** [review]
Nice. Added a naming comment/nit.

**Pull request merged**